proposals/p3762.md

```diff
@@ -0,0 +1,828 @@
+        compilation of libraries, which in turn supports scaling compilation.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   `extern` is chosen for consistency with C++, and carries a similar --
+        albeit slightly different -- meaning.
+-   [Principle: Prefer providing only one way to do a given thing](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/one_way.md)
+    -   Setting _requirements_ for whether a keyword belongs on a forward
+        declaration or with the definition, instead of making places _optional_,
+        supports developers making conclusions based on which keywords they see
+        -- either by presence or absence.
+
+## Alternatives considered
+
+### Other modifier keyword merging approaches
+
+There has been intermittent discussion about which modifiers to allow or require
+on forward declarations versus definitions. There are advantages and
+disadvantages about redundancy and being able to copy-paste declarations. There
+might be strict requirements for some modifiers to be present in order to
+correctly use a forward declaration.
+
+This proposal suggests a partial decision here, at least for a reasonable
+starting point that we can implement towards. This will likely evolve in future
+proposals, particularly as more keywords are added. However, this still offers a
+baseline.
+
+The trade-offs we consider are:
+
+-   Consistency in when a modifier keyword is expected (if applicable) is
+    valuable.
+    -   Adding a keyword to an entity with a separate definition may require
+        adding the keyword to the forward declaration, the separate definition,
+        or both. It is disallowed where not required to be added.
+        -   For example, `base` is added to `class C { ... }`, and disallowed on
+            `class C;`.
+    -   Although we could make keywords optional where it would not affect
+        semantics, we prefer for the presence or absence of a keyword to carry a
+        clear meaning.
+        -   For example, if `base` were optional to allow `base class C;`, then
+            an adjacent `class D;` lends itself to being incorrectly interpreted
+            as meaning "`D` is not a base class" when it actually means "`D` may
+            or may not be a base class".
+-   Access control has a certain necessity for consistency, so that a consumer
+    of a forward declaration would still be allowed to use the definition if
+    refactored.
+    -   We could require similar consistency on `extern`, and should have more
+        nuanced rules if the access control rules go beyond public and
+        `private`. For example, a package-private type shouldn't be allowed to
+        be made public through an `extern` declaration; but package-private
+        isn't actually part of the language right now.
```

Comment:
This argument does not apply to type-scope access control, where we are deviating from C++. I think it's worth saying something about that -- presumably the copy-paste argument given below for functions is the relevant concern here, but I believe this also applies to nested classes (and any other kinds of nested entities, such as interfaces):
```
class A {
  private class B;
}
// `private` is required here
private class A.B {
  // ...
}
```